src/devnet-utils.ts

```diff
@@ -127,17 +130,61 @@ export class DevnetUtils implements Devnet {
 
     public async dump(path: string) {
         return this.withErrorHandler<void>(async () => {
-            await axios.post(`${this.endpoint}/dump`, {
-                path
-            });
+            // make sure the destination directory exists
+            fs.mkdirSync(fspath.dirname(path), { recursive: true });
+
+            // make sure the destination file is always deleted
+            if (fs.existsSync(path)) {
+                fs.unlinkSync(path);
+            }
+
+            // the server will reply immediately and dumping is done in a background thread
+            await axios.post(`${this.endpoint}/dump`,
+                { path },
+                { timeout: 20000 } // timeout is required because the server can die
+            );
+            await this.waitUntilSaveFinished(path);
         }, "Request failed. Make sure your network has the /dump endpoint");
     }
 
+    async waitUntilSaveFinished(path:string): Promise<void> {
+        const maxWaitMillis = 20_000;
+        const startTime = new Date().getTime();
+
+        async function waitWithTimeout(waitMillis:number): Promise<void> {
+            if ((new Date().getTime() - startTime) > maxWaitMillis) {
+                throw new HardhatPluginError(PLUGIN_NAME, "devnet.dump() timed out");
+            }
+            await sleep(waitMillis);
+        }
+
+        // wait until file actually exists
+        while (!fs.existsSync(path) || fs.statSync(path).size === 0) {
+            await waitWithTimeout(50);
+        }
+
+        // and wait until file size finishes changing
+        let size = 0, lastSize = 0;
+        while ((size = fs.statSync(path).size) !== lastSize) {
+            lastSize = size;
+            await waitWithTimeout(50);
+        }
+    }
+
+    /**
+     * Loads devnet state from disk
+     */
     public async load(path: string) {
         return this.withErrorHandler<void>(async () => {
-            await axios.post(`${this.endpoint}/load`, {
-                path
-            });
+            if (!fs.existsSync(path)) {
```

Comment:
What if someone is running their devnet in a different environment, e.g. a container or something remote? This check should be added server-side.